File: examples/leak-checker.c
/// @file leak-checker.c
/// @brief Temporary example program to check for memory leaks.
#include "acquire.zarr.h"

#include <math.h>
#include <stdbool.h>
#include <stdio.h>
#include <stdlib.h>

int
main()
{
    const size_t width = 1920;
    const size_t height = 1080;

    // Configure compression
    ZarrCompressionSettings compression = {
        .compressor = ZarrCompressor_Blosc1,
        .codec = ZarrCompressionCodec_BloscZstd,
        .level = 1,
        .shuffle = 1,
    };

    // Configure stream settings
    ZarrStreamSettings settings = {
        .store_path = "output_leak_check.zarr",
        .s3_settings = NULL,
        .compression_settings = &compression,
        .data_type = ZarrDataType_uint16,
        .version = ZarrVersion_3,
        .multiscale = true
    };

    // Set up dimensions (t, y, x)
    ZarrStreamSettings_create_dimension_array(&settings, 3);

    settings.dimensions[0] = (ZarrDimensionProperties){
        .name = "t",
        .type = ZarrDimensionType_Time,
        .array_size_px = 0, // Unlimited
        .chunk_size_px = 32,
        .shard_size_chunks = 1,
    };

    settings.dimensions[1] = (ZarrDimensionProperties){
        .name = "y",
        .type = ZarrDimensionType_Space,
        .array_size_px = height,
        .chunk_size_px = (height + 128 + 1) / 128,
        .shard_size_chunks = 84,
    };

    settings.dimensions[2] = (ZarrDimensionProperties){
        .name = "x",
        .type = ZarrDimensionType_Space,
        .array_size_px = width,
        .chunk_size_px = 128,
        .shard_size_chunks = (width + 128 + 1) / 128,
    };

    // Create stream
    ZarrStream* stream = ZarrStream_create(&settings);
    // Free Dimension array
    ZarrStreamSettings_destroy_dimension_array(&settings);

    if (!stream) {
        fprintf(stderr, "Failed to create stream\n");
        return 1;
    }

    // Create sample data
    uint16_t* frame = (uint16_t*)malloc(width * height * sizeof(uint16_t));

    // Write frames
    size_t bytes_written;
    for (int t = 0; t < 1000; t++) {
        // Fill frame with a moving diagonal pattern
        for (size_t y = 0; y < height; y++) {
            for (size_t x = 0; x < width; x++) {
                // Create a diagonal pattern that moves with time
                // and varies intensity based on position
                int diagonal = (x + y + t * 8) % 32;

                // Create intensity variation
                uint16_t intensity;
                if (diagonal < 16) {
                    intensity = (uint16_t)((diagonal * 4096)); // Ramp up
                } else {
                    intensity = (uint16_t)((31 - diagonal) * 4096); // Ramp down
                }

                // Add some circular features
                int centerX = width / 2;
                int centerY = height / 2;
                int dx = x - centerX;
                int dy = y - centerY;
                int radius = (int)sqrt(dx * dx + dy * dy);

                // Modulate the pattern with concentric circles
                if (radius % 16 < 8) {
                    intensity = (uint16_t)(intensity * 0.7);
                }

                frame[y * width + x] = intensity;
            }
        }

        printf("Appending frame %d\n", t + 1);
        ZarrStatusCode status = ZarrStream_append(
          stream, frame, width * height * sizeof(uint16_t), &bytes_written);

        if (status != ZarrStatusCode_Success) {
            fprintf(stderr,
                    "Failed to append frame: %s\n",
                    Zarr_get_status_message(status));
            break;
        }
    }

    // Cleanup
    free(frame);
    ZarrStream_destroy(stream);
    return 0;
}

File: examples/python/leak-checker.py
# Zarr V3 with LZ4 compression to filesystem
import numpy as np
from acquire_zarr import (
    StreamSettings, ZarrStream, Dimension, DimensionType, ZarrVersion,
    DataType, Compressor, CompressionCodec, CompressionSettings
)

frame_width = 14192
frame_height = 10640
chunk_width = chunk_height = 128

shard_height = (frame_height + chunk_height + 1) // chunk_height
shard_width = (frame_width + chunk_width + 1) // chunk_width

chunk_planes = 32


def make_sample_data():
    return np.random.randint(
        0, 65535,
        (1, frame_height, frame_width),
        dtype=np.uint16
    )

def main():
    settings = StreamSettings()

    # Configure compression
    settings.compression = CompressionSettings(
        compressor=Compressor.BLOSC1,
        codec=CompressionCodec.BLOSC_LZ4,
        level=1,
        shuffle=1,
    )

    # Configure 5D array (t, c, z, y, x)
    settings.dimensions.extend([
        Dimension(
            name="t",
            kind=DimensionType.TIME,
            array_size_px=0,
            chunk_size_px=chunk_planes,
            shard_size_chunks=1,
        ),
        Dimension(
            name="y",
            kind=DimensionType.SPACE,
            array_size_px=frame_height,
            chunk_size_px=chunk_height,
            shard_size_chunks=shard_height,
        ),
        Dimension(
            name="x",
            kind=DimensionType.SPACE,
            array_size_px=frame_width,
            chunk_size_px=chunk_width,
            shard_size_chunks=shard_width,
        ),
    ])

    settings.store_path = "leak-checker.zarr"
    settings.version = ZarrVersion.V3
    settings.data_type = DataType.UINT16
    settings.multiscale = True

    # Create stream
    stream = ZarrStream(settings)

    # Write sample data
    frame = make_sample_data()
    for frame_id in range(1000):
        print("Appending frame", frame_id + 1)
        stream.append(frame)


if __name__ == "__main__":
    main()
